### PR TITLE
Clean Powerline; Add Dir Support to Clean_Files

### PR DESCRIPTION
**Pull request by davidpfarrell**
##  Adds full-directory support for clean_files.txt

_lint_clean_files.sh_
```
# Pull list of checkable files from clean_files.txt
#  - Folder references are allowed
#  - Empty lines are ignored
#  - Comment lines (#) are ignored
mapfile -t FILES < <(
	cat clean_files.txt \
		| grep -v -E '^\s*$' \
		| grep -v -E '^\s*#' \
		| xargs -n1 -I{} find "{}" -type f
)

BASH_IT='' pre-commit run --files "${FILES[@]}"
```

## Cleans Primary Powerline Theme
To test the directory support, we add `themes/powerline` to the file.
This PR incorporates all the auto-fixes and recommended changes resulting from executing the pre-commit command locally:
```
BASH_IT='' pre-commit run --files $(cat clean_files.txt | xargs -n1 -I{} find {} -type f);
```

NOTE: Using the `"Hide whitespace changes"` features on the `"Files Changed"` tab will make the PR much easier to review.

[edit] Updated local pre-commit run to include prefix `BASH_IT=''` which is needed to suss out source includes that need a `"# shellcheck source"` directive.

### CODE REVIEW

File: themes/powerline/powerline.base.bash
Comment (davidpfarrell) [inline themes/powerline/powerline.base.bash:201]: This is the most complicated change in the PR.
TIL about the `read` function.
Testing locally confirmed that, if you create the `local` variable first, then the read operation will populate the existing variable (vs creating a new, global variable)

**Comment by davidpfarrell** [comment]
TIL that when running the pre-commit checks locally, its best to clear out the `BASH_IT` variable to suss out source includes that need a `"# shellcheck source"` directive.

```
BASH_IT='' pre-commit run --files $(cat clean_files.txt | xargs -n1 -I{} find {} -type f);
```

File: clean_files.txt
Comment (nwinkler) [inline clean_files.txt:1]: Maybe change this to the directory name as well? Remove the file name?

**Review by nwinkler — COMMENTED** [review]
One minor comment/question, please see inline.

Thanks for doing this!

Does it make sense to add the `BASH_IT=''` to the documentation to show how to run `pre-commit` locally? Since the command is getting quite long with all the file/directory parsing, wouldn't it be good to add a script to run `pre-commit` locally?

File: clean_files.txt
Comment (davidpfarrell) [reply clean_files.txt]: Yeah good idea plus its an easy win for this theme since it only contains the one file anyways.

File: docs/make.bat
Comment (davidpfarrell) [inline docs/make.bat:35]: I got confused for a bit, but I *think* what is happening here is that the line-endings for the respository-version of the file were changed after I  : 
* Added ".bat" extension to .gitattributes:
```
*.bat text eol=crlf
```
* Normalized the file:
```
git add --renormalize docs/make.bat 
```

If you hide whitespace changes then this file will hide, but other than that I 'm  not sure how to confirm if the *correct* version of the file will end up in the repo.

**Comment by davidpfarrell** [comment]
> Does it make sense to add the `BASH_IT=''` to the documentation to show how to run `pre-commit` locally? Since the command is getting quite long with all the file/directory parsing, wouldn't it be good to add a script to run `pre-commit` locally?

For running against the clean_files list, I moved the logic into a script, `lint_clean_files.sh` (open to better names) and also modified the travis file to use it.

There's not a great way to unset the BASH_IT variable as par of git's standard per-commit invocation.

But maybe a small section in the docs for using `BASH_IT='' pre-commit` could be useful for testing local changes before committing.

**Comment by NoahGorny** [comment]
Wow @davidpfarrell , great work!
I like the idea of the script and the changes you made, and good job on fixing the broken file. I think it got in as my branch was not rebased when I merged the docs PR.

I think those changes should be in a different PR than this. We should merge the changes and then this PR.

About the docs- please put the sections in `docs/development.rst` and not in `development.md` which will soon be deleted. You can also view the docs in https://bash-it.readthedocs.io/en/latest/

I will give you guys maintainer status in readthedocs.org so you will be able to manage the docs as well

**Comment by davidpfarrell** [comment]
> I think those changes should be in a different PR than this. We should merge the changes and then this PR.

Indeed this PR got away from me and contain(ed) more disparate changes that I like.

As a result, I have moved some of the changes into 3 other PRs:

* Adds better support for .bat files; Fixes docs/make.bat (#1700) (MERGED)
* Moves lint logic into lint_clean_files.sh (#1701) (NEEDS TO BE MERGED BEFORE THIS ONE)
* Adds manual pre-commit usage to docs/development.rst (#1702) (NOT A BLOCKER)

With those PRs moved out, these PR reduces to 5 files, and once #1701 is merged, this PR will reduce to 4 files.

**Comment by NoahGorny** [comment]
Please rebase your PR on master in order to allow us to see the correct diff
I also see you added and extended the scripts here, I will look after you will update this PR

**Comment by davidpfarrell** [comment]
> Please rebase your PR on master in order to allow us to see the correct diff
> I also see you added and extended the scripts here, I will look after you will update this PR

Hi @NoahGorny ,

I merged master, resolved conflicts, then pushed, but then decided to squash since the original commit history was so messy.

So now there's only one commit to be merged, even though the actual PR diff hasn't changed.

Thanks and lemme know if you have an questions/observations!

File: lint_clean_files.sh
Comment (NoahGorny) [inline lint_clean_files.sh:16]: You should add a comment here explaining why you do this

**Review by NoahGorny — APPROVED** [review]
Just minor nitpick, change it and it's a go from me!

File: lint_clean_files.sh
Comment (davidpfarrell) [reply lint_clean_files.sh:23]: Done, good suggestion, thanks !

**Comment by NoahGorny** [comment]
Great job @davidpfarrell !

**Pull request merged**
